### PR TITLE
Metamorph provision support

### PR DESCRIPTION
**Pull request by Jurisak**
This commit is initial implementation. It will be extended.

+ Included intial tests
- without documentation

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -7,5 +7,7 @@ python:
 install:
   - python setup.py install
   - pip install requests
+  - pip install GitPython
+  - pip install PyYaml
```
Comment (robnester-rh) [inline .travis.yml]: I'd suggest using a requirements.txt file containing the requirements.

File: docs/source/index.rst [hunk 2]
```diff
@@ -55,4 +55,22 @@ How to run this plugin:
 and how to run resultsdb ansible module:
 
 * with job_names provided: ``ansible <host> -m resultsdb -a "test_tier=1 nvr=name-version-release job_names=first-job,second-job resultsdb_api_url=resultsdb-url"``
-* without job_names: ``ansible <host> -m resultsdb -a "test_tier=1 nvr=name-version-release resultsdb_api_url=resultsdb-url"``
\ No newline at end of file
+* without job_names: ``ansible <host> -m resultsdb -a "test_tier=1 nvr=name-version-release resultsdb_api_url=resultsdb-url"``
+
+Provision
++++++++++
+Provision plugin purpose is to create topology files. These files will be handled by linch pin tool.
```
Comment (robnester-rh) [inline docs/source/index.rst]: Reword to:
The purpose of the provision plugin is to create topology files.

File: docs/source/index.rst [hunk 3]
```diff
@@ -55,4 +55,22 @@ How to run this plugin:
...
+This plugin created two files:
```
Comment (robnester-rh) [inline docs/source/index.rst]: nit: correct to creates

File: metamorph/plugins/morph_provision.py [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+            raise ProvisionException("Error during cloning git repository {0} with detail: {1}".format(git_repo,
+                                                                                                       detail))
+
+    def setup_topology_by_osp_config(self, osp_config_path):
+        """
+        Method for actualizing topology values from osp config
+
+        :param osp_config_path -- Path to osp config in cloned repository
+        """
+        with open(osp_config_path) as osp_config:
```
Comment (robnester-rh) [inline metamorph/plugins/morph_provision.py]: I'd wrap this in a try to ensure the path exists.

File: metamorph/plugins/morph_provision.py [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+        self.provision_topology['site'] = self.osp_data['sites'][0]['site']
+
+    def setup_topology_by_metadata(self, metadata_path, metadata_location):
+        """
+        Method for actualizing topology values by provided yaml metadata file
+
+        :param metadata_path -- Path to metadata file in cloned repository
+        :param metadata_location -- Dictionary of searched metadata in given metadata file
+        """
+        with open(metadata_path) as metadata_fp:
```
Comment (robnester-rh) [inline metamorph/plugins/morph_provision.py]: Wrap in a try to ensure the path exists.

File: metamorph/plugins/morph_provision.py [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+        """
+        Method for getting metadata value from metadata file by it's path
+
+        :param metadata_source -- extracted metadata from --metadata-file parameter
+        :param metadata_location -- searching metadata location
+        :param metadata_name -- name of metadata searched in given metadata location
+
+        :returns found metadata
+        """
+        if type(metadata_source) is list:
```
Comment (robnester-rh) [inline metamorph/plugins/morph_provision.py]: Just a note:
Generally speaking, I think it's more pythonic to employ duck typing and do something like:
```
if isinstance(metadata_source, list):
```

File: metamorph/plugins/morph_provision.py [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+    def get_git_repo_name(git_repo):
+        """
+        Method for getting git repository name
+
+        :param git_repo -- git repository path
+
+        :returns repository name
+        """
+        if git_repo.endswith('/'):
+            git_repo = git_repo[:-1]
```
Comment (robnester-rh) [inline metamorph/plugins/morph_provision.py]: easier just to do:
```
git_repo = git_repo.rstrip('/')
```
If git_repo ends with '/' it'll strip it. If it doesn't end with '/' git_repo will be unaffected.

File: metamorph/plugins/morph_provision.py [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+        splitted_repo_path = git_repo.split('/')
+        return splitted_repo_path[-1].split('.')[0]
```
Comment (robnester-rh) [inline metamorph/plugins/morph_provision.py]: this whole method could be rewritten as:

```
return git_repo.rstrip('/').split('/')[-1].split('.')[0]
```

File: metamorph/plugins/morph_provision.py [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+
+    :param args -- argparse object
+    """
+    location = dict()
+    for single_mapping in args.metadata_loc:
+        location[single_mapping[0]] = single_mapping[1].split(',')
+    args.metadata_loc = location
+
+
+def write_results(destination, results, json_type=True):
```
Comment (robnester-rh) [inline metamorph/plugins/morph_provision.py]: Wrap in a try in case the destination doesn't exist.

**Review by robnester-rh — COMMENTED** [review]
I would suggest looking at cases where you're opening paths to read or write from and consider wrapping them with a try, in case the path doesn't exist.

**Comment by gsr-shanks** [comment]
Please resolve conflicts and implement review comments.

File: docs/source/index.rst (hunk 3 above)
Comment (Jurisak) [reply docs/source/index.rst]: thanks

File: metamorph/plugins/morph_provision.py (hunk 4 above)
Comment (Jurisak) [reply metamorph/plugins/morph_provision.py]: yes :)

File: metamorph/plugins/morph_provision.py (hunk 5 above)
Comment (Jurisak) [reply metamorph/plugins/morph_provision.py]: ok

File: metamorph/plugins/morph_provision.py (hunk 6 above)
Comment (Jurisak) [reply metamorph/plugins/morph_provision.py]: looks better ;) thanks

File: metamorph/plugins/morph_provision.py (hunk 7 above)
Comment (Jurisak) [reply metamorph/plugins/morph_provision.py]: I did not know about it. Thanks.

File: metamorph/plugins/morph_provision.py (hunk 8 above)
Comment (Jurisak) [reply metamorph/plugins/morph_provision.py]: Yes but I like my version more.

File: metamorph/plugins/morph_provision.py (hunk 9 above)
Comment (Jurisak) [reply metamorph/plugins/morph_provision.py]: sure

**Comment by gsr-shanks** [comment]
Please resolve conflicts

**Comment by Jurisak** [comment]
I don't have time to do it today, but I'll be able to do it tomorrow.

**Comment by Jurisak** [comment]
@gsr-shanks It can be merged now.

**Comment by gsr-shanks** [comment]
@Jurisak are we blocked on https://github.com/CentOS-PaaS-SIG/linch-pin/issues/76 for this to work ?

**Comment by Jurisak** [comment]
@gsr-shanks Yes, it would need a little change. Expect the patch today.

**Comment by Jurisak** [comment]
@gsr-shanks @robnester-rh @Lorquas Guys, how do you see it with merging this PR?

File: .travis.yml [hunk 10]
```diff
@@ -4,7 +4,7 @@ python:
   - "3.5"
   - "3.6"
 install:
-  - python setup.py install
+  - python setup.py install
```
Comment (Lorquas) [inline .travis.yml]: don't need the extra space

File: docs/source/index.rst [hunk 11]
```diff
@@ -67,6 +67,23 @@ and how to run resultsdb ansible module:
 * with job_names provided: ``ansible <host> -m resultsdb -a "test_tier=1 nvr=name-version-release job_names=first-job,second-job resultsdb_api_url=resultsdb-url"``
 * without job_names: ``ansible <host> -m resultsdb -a "test_tier=1 nvr=name-version-release resultsdb_api_url=resultsdb-url"``
 
+Provision
++++++++++
+Provision plugin purpose is to create topology files. These files will be handled by linch pin tool.
```
Comment (Lorquas) [inline docs/source/index.rst]:
```
The provision plugin's purpose is to create topology files. These files will be handled by the [linch pin](https://github.com/CentOS-PaaS-SIG/linch-pin) tool.  
```

File: docs/source/index.rst [hunk 12]
```diff
@@ -67,6 +67,23 @@ and how to run resultsdb ansible module:
...
+This plugin creates two files:
+* topology.json - contains all data needed for provisioning
+* topology_credentials.yaml - contains needed credentials for VM creation
+These two files are created from openstack config file.
```
Comment (Lorquas) [inline docs/source/index.rst]:
```
These two files are created from an opestack config file.
```
also might need to link to an openstack config url or something showing what that is.

File: docs/source/index.rst [hunk 13]
```diff
@@ -67,6 +67,23 @@ and how to run resultsdb ansible module:
...
+
+Provision plugin provides possibility to update topology values from metadata file and metadata location dictionary
```
Comment (Lorquas) [inline docs/source/index.rst]: ~~possibility~~  >>  the ability

~~from metadata file and~~ >> from the metadata file and the

File: docs/source/index.rst [hunk 14]
```diff
@@ -91,3 +108,20 @@ Easiest way to run this plugin is by:
 Extracted data are stored in json file.
 
 
+Provision
```
Comment (Lorquas) [inline docs/source/index.rst]: I would check this entire file, it seems to contain the same information twice!

File: setup.py [hunk 15]
```diff
@@ -8,5 +8,5 @@
     license='',
     author='Jiri Kulda',
     author_email='Kulda12@seznam.cz, jkulda@redhat.com',
-    description='Exhibiting structural change of test metadata.', requires=['ansible', 'requests']
+    description='Exhibiting structural change of test metadata.', requires=['ansible', 'requests', 'GitPython']
```
Comment (Lorquas) [inline setup.py]: the description and requires can be on separate lines

**Review by Lorquas — CHANGES_REQUESTED** [review]
@Jurisak as always, code looks great

Few small nitpicks, and the docs looks it has repeated everything twice.  Looks good to me after these changes.

File: .travis.yml (hunk 10 above)
Comment (Jurisak) [reply .travis.yml]: thanks

**Comment by Jurisak** [comment]
@Lorquas  How do you see it now?

**Review by spbirhade — COMMENTED** [review]
Looks ok

**Pull request merged**